app/controllers/RobotsController.php

```diff
@@ -0,0 +1,357 @@
+                    $this->flash->error($message);
+                }
+            } else {
+                $robot = new Robots([
+                    'name' => $this->request->getPost('name', 'striptags'),
+                    'type' => $this->request->getPost('type', 'striptags'),
+                    'year' => $yearList[$this->request->getPost('year', 'int')],
+                ]);
+
+                $robot->parts = $this->getPartsToInsert($this->request->getPost('partList', 'striptags'));
+
+                if (! $robot->save()) {
+                    $this->flash->error($robot->getMessages());
+                } else {
+                    $this->flash->success("Robot was created successfully");
+
+                    $form->clear();
+                }
+            }
+        }
+
+        $this->view->form = $form;
+    }
+
+    /**
+     * Saves the user from the 'edit' action
+     */
+    public function editAction($id)
+    {
+        $robot = Robots::findFirstById($id);
+
+        $years = range(1900, (int) date('Y'));
+        $yearList = array_combine($years, $years);
+
+        $partNames = [];
+
+        /** @var Parts $part */
+        foreach ($robot->parts as $part) {
+            $partNames[] = $part->name;
+        }
+
+        $partNamesAsString = implode(',', $partNames);
+
+        $options = [
+            'years' => $yearList,
+            'edit' => true,
+            'partNames' => $partNamesAsString,
+        ];
+
+        if (! $robot) {
+            $this->flash->error("Robot was not found");
+
+            return $this->dispatcher->forward([
+                'action' => 'index'
+            ]);
+        }
+
+        if ($this->request->isPost()) {
+            $robot->assign([
+                'name' => $this->request->getPost('name', 'striptags'),
+                'type' => $this->request->getPost('type', 'striptags'),
+                'year' => $yearList[$this->request->getPost('year', 'int')],
+            ]);
+
+            $form = new RobotsForm($robot, $options);
+
+            $this->processRobotParts(
+                $robot,
+                $partNamesAsString,
+                trim($this->request->getPost('partList', 'striptags'))
+            );
+
+            if ($form->isValid($this->request->getPost()) == false) {
+                foreach ($form->getMessages() as $message) {
+                    $this->flash->error($message);
+                }
+            } else {
+                if (! $robot->save()) {
+                    $this->flash->error($robot->getMessages());
+                } else {
+
+                    $this->flash->success("Robot was updated successfully");
+
+                    $form->clear();
+                }
+            }
+        }
+
+        $this->view->robot = $robot;
+
+        $this->view->form = new RobotsForm($robot, $options);
+    }
+
+    protected function processRobotParts($robot, $oldPartNames, $newPartNames)
```

Comment:
Method finds 3 states of the relationship:
* not changed,
* to delete
* to add